File: day1/day1.cpp
#include <fstream>
#include <vector>
using namespace std;

int part1(ifstream &infile) {
  int sum = 0;
  string line;
  while (getline(infile, line)) {
    auto it = line.find_first_of("0123456789");
    int num1 = line[it] - '0';
    it = line.find_last_of("0123456789");
    int num2 = line[it] - '0';
    sum += num1 * 10 + num2;
  }
  return sum;
}

int findFirstNumber(const string &line, const vector<string> &lines) {
  size_t firstPos = string::npos;
  int firstNum = -1;

  for (int i = 0; i < lines.size(); i++) {
    size_t pos = line.find(lines[i]);
    if (pos != string::npos && (firstPos == string::npos || pos < firstPos)) {
      firstPos = pos;
      firstNum = i + 1;
    }
  }

  size_t digitPos = line.find_first_of("0123456789");
  if (digitPos != string::npos &&
      (firstPos == string::npos || digitPos < firstPos)) {
    return line[digitPos] - '0';
  }
  return firstNum;
}

int findLastNumber(const string &line, const vector<string> &lines) {
  size_t lastPos = string::npos;
  int lastNum = -1;

  for (int i = 0; i < lines.size(); i++) {
    size_t pos = line.rfind(lines[i]);
    if (pos != string::npos && (lastPos == string::npos || pos > lastPos)) {
      lastPos = pos;
      lastNum = i + 1;
    }
  }

  size_t digitPos = line.find_last_of("0123456789");
  if (digitPos != string::npos &&
      (lastPos == string::npos || digitPos > lastPos)) {
    return line[digitPos] - '0';
  }
  return lastNum;
}

int part2(ifstream &infile) {
  vector<string> lines = {"one", "two",   "three", "four", "five",
                          "six", "seven", "eight", "nine"};
  int sum = 0;
  string line;
  while (getline(infile, line)) {
    int num1 = findFirstNumber(line, lines);
    int num2 = findLastNumber(line, lines);

    if (num1 != -1 && num2 != -1) {
      sum += num1 * 10 + num2;
    }
  }
  return sum;
}
int main() {
  ifstream infile("input");
  printf("Part 1: %d\n", part1(infile));
  infile = ifstream("input");
  printf("Part 2: %d\n", part2(infile));
}

File: day1/day1.py
inp = [l.strip() for l in open('input')]
nums = []

m = {"one": 1, "two": 2, "three": 3, "four": 4, "five": 5,
     "six": 6, "seven": 7, "eight": 8, "nine": 9}


# Part 1
for word in inp:
    res = ""
    for char in word:
        if char.isdigit():
            res += char

    res = res[0] + res[-1]
    nums.append(int(res))

print(sum(nums))

# Part 2
nums = []
for word in inp:
    res = ""
    for num in m:
        if num in word:
            word = word.replace(num, num + str(m[num]) + num)
    for char in word:
        if char.isdigit():
            res += char

    res = "".join(res)
    res = res[0] + res[-1]
    nums.append(int(res))


print(sum(nums))
